### PR TITLE
kubelet: remove the --enable-cadvisor-json-endpoints flag

### LINKED ISSUE #68522: Reduce the set of metrics exposed by the kubelet
### Background
In 1.12, the kubelet exposes a number of sources for metrics directly from [cAdvisor](https://github.com/google/cadvisor#cadvisor).  This includes:
 * [cAdvisor prometheus metrics](https://github.com/google/cadvisor/blob/master/docs/storage/prometheus.md#prometheus-metrics) at [`/metrics/cadvisor`](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/server/server.go#L277)
 * [cAdvisor v1 Json API](https://github.com/google/cadvisor/blob/master/info/v1/container.go#L126) at [`/stats/`, `/stats/container`, `/stats/{podName}/{containerName}`, and `/stats/{namespace}/{podName}/{uid}/{containerName}`](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/server/stats/handler.go#L111)
 * [cAdvisor machine info](https://github.com/google/cadvisor/blob/master/info/v1/machine.go#L159) at [`/spec`](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/server/server.go#L291)

The kubelet also exposes the [summary API](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/apis/stats/v1alpha1/types.go#L24), which is not exposed directly by cAdvisor, but queries cAdvisor as one of its sources for metrics.

The [Monitoring Architecture](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/instrumentation/monitoring_architecture.md) documentation describes the path for "core" metrics, and for "monitoring" metrics.  The [Core Metrics](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/instrumentation/core-metrics-pipeline.md#core-metrics-in-kubelet) proposal describes the set of metrics that we consider core, and their uses.  The motivation for the split architecture is:
 * To minimize the performance impact of stats collection for core metrics, allowing these to be collected more frequently
 * To make the monitoring pipeline replaceable, and extensible.

### Current kubelet metrics that are not included in core metrics
 * Pod and Node-level Network Metrics
 * Persistent Volume Metrics
 * Container-level (Nvidia) GPU Metrics
 * Node-Level RLimit Metrics
 * Misc Memory Metrics (e.g. PageFaults)
 * Container, Pod, and Node-level Inode metrics (for ephemeral storage)
 * Container, Pod, and Node-level DiskIO metrics (from cAdvisor)

Deprecating and removing the Summary API will require out-of-tree sources for each of these metrics.  "Direct" cAdvisor endpoints are not often used, and have even been broken for multiple releases (https://github.com/kubernetes/kubernetes/pull/62544) without anyone raising an issue.

### Working Items
 * [x] [1.13] Introduce Kubelet `pod-resources` grpc endpoint; KEP: https://github.com/kubernetes/community/pull/2454
 * [x] [1.14] Introduce Kubelet Resource Metrics API
 * [x] [1.15] Deprecate the "direct" cAdvisor API endpoints by adding and deprecating a `--enable-cadvisor-json-endpoints` flag
 * [x] [1.18] Default the `--enable-cadvisor-json-endpoints` flag to disabled
 * [ ] [1.21] Remove the `--enable-cadvisor-json-endpoints` flag
 * [ ] [1.21] Transition Monitoring Server to Kubelet Resource Metrics API ([requires 3 versions skew](https://github.com/kubernetes/kubernetes/pull/67829#issuecomment-416873857))
 * [ ] [TBD] Propose out-of-tree replacements for kubelet monitoring endpoints
 * [ ] [TBD] Deprecate the Summary API and cAdvisor prometheus endoints by adding and deprecating a `--enable-container-monitoring-endpoints` flag
 * [ ] [TBD+2] Remove "direct" cAdvisor API endpoints
 * [ ] [TBD+2] Default the `--enable-container-monitoring-endpoints` flag to disabled
 * [ ] [TBD+4] Remove the Summary API, cAdvisor prometheus metrics and remove the `--enable-container-monitoring-endpoints` flag.

### Open Questions
 * Should the kubelet be a source for any monitoring metrics?
   * For example, metrics about the kubelet itself, or DiskIO metrics for empty-dir volumes (which are "owned" by the kubelet).
 * What will provide the metrics listed above, now that the kubelet no longer does?
   * cAdvisor can provide Network, RLimit, Misc Memory metrics, Inode metrics, and DiskIO metrics.
     * cAdvisor only works for some runtimes, but is a drop-in replacement for "direct" cAdvisor API endpoints
   * Container Runtimes can be a source for container-level Memory, Inode, Network and DiskIO metrics.
   * NVidia GPU metrics provided by a daemonset published by NVidia
   * No source for Persistent Volume metrics?

/sig node
/sig instrumentation
/kind feature
/priority important-longterm
cc @kubernetes/sig-node-proposals @kubernetes/sig-instrumentation-misc

### PR DESCRIPTION
**Pull request by utkarsh2102**
Hello 👋🏻 

**What type of PR is this?**

/kind cleanup
/kind deprecation
(hopefully, these are relevant ones here!)

**What this PR does / why we need it**:

This partially fixes #68522, with one of its tasks to remove the already deprecated `--enable-cadvisor-json-endpoints` flag.

**Special notes for your reviewer**:

I am not really versed with Go so in case things look out of order that's likely the reason behind this! :)
Irrespective of this, please give me a shout and I'll be happy to do the fixes at the earliest!

**Does this PR introduce a user-facing change?**:

```release-note
kubelet: remove the already deprecated --enable-cadvisor-json-endpoints flag.
```

Signed-off-by: Utkarsh Gupta <<utkarsh@debian.org>>

---

/cc @dims

### CODE REVIEW

**Comment by utkarsh2102** [comment]
> It may take a couple minutes for the CLA signature to be fully registered; after that, please reply here with a new comment and we'll verify.

CLA has been duly signed! 🌮 

> To complete the pull request process, please assign sjenning after the PR has been reviewed.
> You can assign the PR to them by writing /assign @sjenning in a comment when ready.

/assign @sjenning 
/assign @dims

File: pkg/kubelet/cadvisor/cadvisor_linux.go [hunk 1]
```diff
@@ -31,12 +31,6 @@ import (
 	_ "github.com/google/cadvisor/container/crio/install"
 	_ "github.com/google/cadvisor/container/systemd/install"
 
-	// Register cloud info providers.
```
Comment (dims) [inline pkg/kubelet/cadvisor/cadvisor_linux.go:34]: @utkarsh2102 please leave this section in for now.

File: pkg/kubelet/server/stats/handler.go [hunk 2]
```diff
@@ -113,29 +113,22 @@ type handler struct {
...
-	}
-
-	if enableCAdvisorJSONEndpoints {
-		endpoints = append(endpoints,
-			endpoint{"", h.handleStats},
-			endpoint{"/container", h.handleSystemContainer},
-			endpoint{"/{podName}/{containerName}", h.handlePodContainer},
-			endpoint{"/{namespace}/{podName}/{uid}/{containerName}", h.handlePodContainer},
-		)
+		{"/container", h.handleSystemContainer},
```
Comment (dims) [inline pkg/kubelet/server/stats/handler.go:129]: enableCAdvisorJSONEndpoints was off by default, so we can remove lines 129,130,131

File: pkg/kubelet/cadvisor/cadvisor_linux.go (hunk 1 above)
Comment (utkarsh2102) [reply pkg/kubelet/cadvisor/cadvisor_linux.go]: Re-added! 👍🏻

File: pkg/kubelet/server/stats/handler.go (hunk 2 above)
Comment (utkarsh2102) [reply pkg/kubelet/server/stats/handler.go]: Aaah! Removed, thanks!

File: pkg/kubelet/server/server.go [hunk 3]
```diff
@@ -393,17 +391,15 @@ func (s *Server) InstallDefaultHandlers(enableCAdvisorJSONEndpoints bool) {
...
-		ws.
-			Path(specPath).
-			Produces(restful.MIME_JSON)
-		ws.Route(ws.GET("").
-			To(s.getSpec).
-			Operation("getSpec").
-			Writes(cadvisorapi.MachineInfo{}))
-		s.restfulCont.Add(ws)
-	}
+	ws := new(restful.WebService)
```
Comment (utkarsh2102) [inline pkg/kubelet/server/server.go:394]: This is as interesting as it gets. I get:
```
# k8s.io/kubernetes/pkg/kubelet/server
pkg/kubelet/server/server.go:394:5: no new variables on left side of :=
# k8s.io/kubernetes/pkg/kubelet/server
vet: pkg/kubelet/server/server.go:394:5: no new variables on left side of :=
make[1]: *** [Makefile:388: vet] Error 1
```
and a bunch of other failures are probably related as well.

Interestingly, all I did was revert some part of #78504, where this was exactly like this. Likely I am missing somehting (not sure what though!?)

**Comment by SergeyKanzhelev** [comment]
/assign @bobbypage 

@bobbypage can you take a look?

**Comment by neolit123** [comment]
> Remove the already deprecated --enable-cadvisor-json-endpoints flag.

please prefix the release note with `kubelet: ` or mention that this is a kubelet flag.

File: pkg/kubelet/server/stats/handler.go [hunk 4]
```diff
@@ -113,31 +113,21 @@ type handler struct {
...
+	}{
+		{"", h.handleStats},
 		{"/summary", h.handleSummary},
 	}
 
-	if enableCAdvisorJSONEndpoints {
-		endpoints = append(endpoints,
-			endpoint{"", h.handleStats},
-			endpoint{"/container", h.handleSystemContainer},
-			endpoint{"/{podName}/{containerName}", h.handlePodContainer},
```
Comment (SergeyKanzhelev) [inline pkg/kubelet/server/stats/handler.go:136]: this is mentioned in pkg/kubelet/server/auth_test.go Perhaps clean it up as well

Comment (utkarsh2102) [reply pkg/kubelet/server/stats/handler.go:136]: Aye, thanks, did that. There are some test failures that I need a little help with though :)

File: pkg/kubelet/server/auth_test.go [hunk 5]
```diff
@@ -138,8 +138,6 @@ func AuthzTestCases() []AuthzTestCase {
 		"/stats/":          "stats",
 		"/stats/container": "stats",
```
Comment (SergeyKanzhelev) [inline pkg/kubelet/server/auth_test.go:139]: this one too, right?

File: pkg/kubelet/server/stats/handler.go (hunk 4 above)
Comment (SergeyKanzhelev) [reply pkg/kubelet/server/stats/handler.go:136]: Test failure: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/95763/pull-kubernetes-bazel-test/1319488038476189696

Failure:

```
--- FAIL: TestContainerInfo (0.02s)
    server_test.go:389: received invalid json data: json: cannot unmarshal number into Go value of type v1.ContainerInfo
I1023 04:30:17.123268  228861 server.go:398] Adding debug handlers to kubelet server.
```

So the test `func TestContainerInfo(t *testing.T) {` validates that this endpoint (`/{podName}/{containerName}`) returns json. Endpoint constructed here:

```
resp, err := http.Get(fw.testHTTPServer.URL + fmt.Sprintf("/stats/%v/%v", podID, expectedContainerName))
```

 It fails with `json: cannot unmarshal number into Go value of type v1.ContainerInfo`. So I'd guess that when endpoint is missing, `404` is returned and this response cannot be parsed as a json. So ultimately just remove the test.

I'm not sure if the removal of these endpoint is what expected, let me take a look at the original issue

**Comment by dims** [comment]
@utkarsh2102 please see the 2 failing tests (click on details and check the "Raw build-log.txt" link)

**Comment by SergeyKanzhelev** [comment]
/hold

See comment: https://github.com/kubernetes/kubernetes/issues/68522#issuecomment-715462010. 

@utkarsh2102 thank you for initiating this work. Are you interested to go thru the enhancement process? It is likely will be for 1.21, unless it would be accepted quite unanimously and we can get an exception. If not, I'd suggest to close this PR for now (sorry for this experience, for the next PR I encourage discuss contribution in the issue first. Not all issues are up to date).

**Comment by utkarsh2102** [comment]
Hey @SergeyKanzhelev,

> @utkarsh2102 thank you for initiating this work. Are you interested to go thru the enhancement process? It is likely will be for 1.21, unless it would be accepted quite unanimously and we can get an exception. If not, I'd suggest to close this PR for now (sorry for this experience, for the next PR I encourage discuss contribution in the issue first. Not all issues are up to date).

Aah. Actually @dims is helping me get started contributing to Kubernetes and we thought this would be a nice way to start things off :)

But that's alright, we can always revisit this later when time's good; thanks for your reviews & help! 😄

**Comment by SergeyKanzhelev** [comment]
@dims @utkarsh2102 I'm actually not against the deprecation. And it is not the best experience for contributions that the issue is not up to date =(. I want to close this PR based on the comment in the issue. But we can re-open if we will get an agreement on deprecation. @dims do you want to bring it to sig node meeting tomorrow?

/close

**Pull request closed without merging**